src/RazorSdk/Targets/Microsoft.NET.Sdk.Razor.JSModules.targets

```diff
@@ -0,0 +1,325 @@
+    * For web applications (Blazor webassembly and Blazor server) the main project is responsible for producing a bundle <project-name>.modules.json that will contain the list of modules to load
+  * For packing in razor class libraries, the JsModule files will get processed and added as static web assets to the pack.
+
+  Integration with static web assets:
+  * The JS modules will be added as regular static web assets to participate in the pipeline.
+    * Generated JS module files will have a unique extension '.module.lib.js' that will be used by the pipeline to identify them as such unless otherwise identified.
+  * In razor class libraries these generated files will be packaged normally as part of the static web assets process and will be accessible at <<StaticWebAssetsBasePath>>/<<RelativePath>>.
+    * Using the extension is useful as it allows for third party tooling to do alternative processing in an easy way, these files will be
+      removed off from the list of static web assets when the default bundling is enabled, so they won't show up in the final output.
+  -->
+
+  <UsingTask TaskName="Microsoft.AspNetCore.Razor.Tasks.GenerateJsModuleManifest" AssemblyFile="$(RazorSdkBuildTasksAssembly)" />
+
+  <PropertyGroup>
+    <GenerateJSModuleManifest Condition="$(GenerateJSModuleManifest) == ''">true</GenerateJSModuleManifest>
+    <ResolveStaticWebAssetsInputsDependsOn>
+      $(ResolveStaticWebAssetsInputsDependsOn);
+      _AddJsModulesToStaticWebAssets;
+    </ResolveStaticWebAssetsInputsDependsOn>
+
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(GenerateJSModuleManifest)' == 'true'">
+    <StaticWebAssetsPrepareForRunDependsOn>
+      $(StaticWebAssetsPrepareForRunDependsOn);
+      _GenerateBuildJsModuleManifest;
+    </StaticWebAssetsPrepareForRunDependsOn>
+
+    <StaticWebAssetsPrepareForPublishDependsOn>
+      $(StaticWebAssetsPrepareForPublishDependsOn);
+      _GeneratePublishJsModuleManifest;
+    </StaticWebAssetsPrepareForPublishDependsOn>
+
+    <ResolveStaticWebAssetsInputsDependsOn>
+      $(ResolveStaticWebAssetsInputsDependsOn);
+      _AddBuildJsModuleManifest;
+    </ResolveStaticWebAssetsInputsDependsOn>
+
+    <ResolvePublishStaticWebAssetsDependsOn>
+      $(ResolvePublishStaticWebAssetsDependsOn);
+      _AddPublishJsModuleManifest
+    </ResolvePublishStaticWebAssetsDependsOn>
+
+  </PropertyGroup>
+
+  <Target Name="_ResolveJsModuleInputs" DependsOnTargets="ResolveProjectStaticWebAssets">
+
+    <ItemGroup>
+      <_JSModuleCandidates Include="@(StaticWebAsset)" Condition="'%(SourceType)' == 'Discovered'" />
+    </ItemGroup>
+
+    <!-- This runs after we've discovered the original set of files, we already found the assets in the wwwroot, and here we just need
+         to identify them and correctly clasify them. Modules from other projects or packages will already be correctly tagged when we
+         retrieve them.
+    -->
```

Comment:
I have strong reservations about us hard-coding filename conventions like `Pages/`, `Shared/`, `Components/`, etc. into the SDK. These are arbitrary details of the project templates and not part of the framework. I personally would not have folder names like those in any project I created (since I would group things by feature, not by file type), and most likely teams working in other languages would have different names for anything they add.

If we need to have a filter rule, can we force the developer to declare it? Something like:

```xml
<ItemGroup>
    <JavaScriptModule Include="MyApp/**/*.js" />
</ItemGroup>
```

... or something like that?

Since it's not obvious that the SDK is going to bundle and publicly ship arbitrary `.js` files from the project, it's ideal for developers to have to declare this explicitly and not have a default. But if there is a default, it definitely shouldn't include predefined strings like `Pages`, `Shared`, etc.

Reply:
Hmm, Pages, Shared, Views, etc. Are at least well-known conventions in the MVC world. Our template ships with a Pages folder that most people use. 

If you want to have a different convention, we can provide a way for users to do that, but I think we should have default conventions that match the patterns we normally follow.

Reply:
I'd love to fix our templates not to have things like `Pages`, `Shared`, etc. On the assumption that we're going to do that (for Blazor at least) it doesn't seem wise to bake them as rules into the SDK.

Reply:
This is done now